### ISSUE TITLE
Crimea location bug

### ISSUE THREAD

**Issue by gerasimvol** [nvkelso/natural-earth-vector#489]
Crimea somehow isn’t part of Ukraine. Please fix it. And we don’t accept any special settings or admin-0 options. It’s not a discussion, it’s a fact.

**Comment by matkoniecz** [comment]
Sadly, Crimea was successfully invaded and taken over by Russia.

See https://www.naturalearthdata.com/downloads/10m-cultural-vectors/10m-admin-0-details/

> Disclaimer

> Natural Earth Vector draws boundaries of sovereign states according to defacto status. We show who actually controls the situation on the ground. For instance, we show China and Taiwan as two separate states. But we show Palestine as part of Israel. Please feel free to mashup our countries and disputed areas themes to match your particular political outlook.

**Comment by matkoniecz** [comment]
See also #401 #391 #390 

Is there some reason why this issues are kept open?

**Comment by gerasimvol** [comment]
See also? https://undocs.org/en/A/RES/68/262
If it was taken illegally and not recognised by any civilised country, how it could be displayed on map as Russia? Are you insane?

**Comment by mmatviyiv** [comment]
@matkoniecz being agreed with "de facto borders" means accepting such invasions.
The smallest thing which the civilised world can do is to keep claiming everywhere that Crimea is Ukraine.
United Nations have already done that, a link is above.

**Comment by nvkelso** [comment]
There are a large number of people following each of the issues. Once my PR is merged the issues will be resolved, and everyone will receive a GH notification for each of the GH issues.

**Comment by anxolerd** [comment]
> being agreed with "de facto borders" means accepting such invasions.

I would say it is not just "accepting" such invasions, but rather supporting them. Next times invaders may consider this "acceptance" as a sign of support.

**Comment by sarnakov** [comment]
I`ve figured out how to fixed it.

I deleted in Russia`s arcs an array [517, 518], and then changed Ukraine`s arcs to 
`"arcs": [
[
	-96, -513, 575, -518, 576, -510, -428, -509, -327,
	-555, -497
],
[517, 518]
]`

**Comment by nvkelso** [comment]
The referenced PR https://github.com/nvkelso/natural-earth-vector/pull/446 has been merged – though v5 remains in development so now the files in main / master branch of GH are accurate but haven't been posted to the naturalearthdata.com website yet.

**Comment by rredfearn32** [comment]
> I`ve figured out how to fix it.
> 
> I deleted in Russia's arcs an array [517, 518], and then changed Ukraine's arcs to `"arcs": [ [ -96, -513, 575, -518, 576, -510, -428, -509, -327, -555, -497 ], [517, 518] ]`

This totally worked for me. I'm using the react-simple-maps plugin, which depends on the natural-earth data. I wrote this function (very messy, I know, I will clean it up later) for updating Ukraine and Russia:

```
const updateCrimea = (geoData: any): any => {
  const geometries = geoData?.objects?.ne_110m_admin_0_countries?.geometries;
  if (geometries) {
    // Remove Crimea from Russia
    const russiaData = geometries.find(
      (i: any) => i.properties.NAME === 'Russia',
    );
    if (russiaData) {
      russiaData.arcs = russiaData.arcs.filter(
        (i: any) => JSON.stringify(i) !== '[[517,518]]',
      );
    }

    // Add Crimea to Ukraine
    const ukraineData = geometries.find(
      (i: any) => i.properties.NAME === 'Ukraine',
    );
    if (ukraineData) {
      ukraineData.type = 'MultiPolygon';
      ukraineData.arcs = [
        [[-96, -513, 575, -518, 576, -510, -428, -509, -327, -555, -497]],
        [[517, 518]],
      ];
    }
  }

  return geoData;
};
```

**Comment by mmatviyiv** [comment]
How many Ukrainian victims do you need in this war to stop even assuming that Crimea can be shown as non-ukrainian land?

**Comment by qis** [comment]
@anxolerd This has been debunked as a Ukrainian false flag operation.

**Comment by webknjaz** [comment]
> @anxolerd This has been debunked as a Ukrainian false flag operation.

Wait a second, why are you supporting the russian nazis? Did they pay you or are you revealing your ideology just now?

**Comment by mmatviyiv** [comment]
If you wonder why do these russians keep calling war politics, because death, tortures, genocide are their state politics.

**Comment by webknjaz** [comment]
> @webknjaz I'm simply stating facts. Why are there politics on this issue in the first place?

It's not "politics". Also, it's not a fact but an amplification of the mordor propaganda. The only ones who spread it are paid bots and russian fascist fanatics.
By amplifying this propaganda, anyone doing this is putting themself in line with the rest of the child rapist russian nazis that invaded my home. By doing this, you not not only passively but actively support this genocide along with the rest of the murders this regime praises.
I see you. I see what you are. Everybody here does too.

Back to the cartography side of the question. Putting the murderous oppressor-demanded borders on a publicly available map helps the propaganda of their regime to come up with excuses for committing more genocide (that's been going on for 8 years, let me remind you, just less severe). And by extension, anyone's hands involved in this less obvious aggressor support are as dirty with the blood of the Ukrainian civilians, young and old, children and women as well as the defenders' blood too.

This is a fact. As simple as it gets. Not politics, just life. A direct result of what people decide here.

**Comment by goliney** [comment]
@nvkelso do you sleep well at night?

**Comment by EddyCMWF** [comment]
Here is some python code for moving the Crimea feature from the Russia multipolygon into the Ukraine multipolygon. Hopefully it is generic enough to use with other shapefiles.

```
import geopandas as gpd
from shapely import Point, MultiPolygon

crimea = {'lat': 45, 'lon': 34}
crimea_point = Point(crimea['lon'], crimea['lat'])

infiles = ['ne_110m_admin_0_map_units.zip']

for infile in infiles:
    outfile = f"{infile.replace('.zip', '')}_crimea"
    country_units = gpd.read_file(infile)
    
    # Get indexes for Ukraine and Russia (needed for updating the DF at end)
    irussia = country_units.index[country_units.ADMIN=='Russia']
    iukraine = country_units.index[country_units.ADMIN=='Ukraine']
    
    # Get Ukraine row and list with singular Ukraine Polygon:
    ukraine = country_units.iloc[iukraine]
    ukraine_polys = [poly for poly in ukraine['geometry']]
    
    # Get the russia row
    russia = country_units.iloc[irussia]
    
    # extract the Russia polygons to a list and append the polygon which contains 
    # The crimea_point to the Ukraine list
    russia_polys = []
    for i, multipoly in enumerate(russia['geometry']):
        for j, poly in enumerate(multipoly.geoms):
            if poly.contains(crimea_point):
                ukraine_polys.append(poly)
            else:
                russia_polys.append(poly)
    
    # Create new Multipolygons with these lists
    ukraine_multi_poly = MultiPolygon(ukraine_polys)
    russia_multi_poly = MultiPolygon(russia_polys)
    
    # Replace in the GPD-DF:
    country_units.loc[iukraine, 'geometry'] = ukraine_multi_poly
    country_units.loc[irussia, 'geometry'] = russia_multi_poly
    
    # Save to output folder
    country_units.to_file(outfile)
```

**Comment by enn-nafnlaus** [comment]
> @matkoniecz being agreed with "de facto borders" means accepting such invasions.

If your standard means "accepting such invasions, then your **standard is wrong.**

I can't begin to tell you how much of a massive pain this horrible invasion-encouraging "standard" is to undo.    Like, for example, trying to get the *acctual* UN-recognized country for Sevastopol.  Here's every single data field:

('SCALERANK', 3)
('NATSCALE', 110)
('LABELRANK', 5.0)
('FEATURECLA', 'Admin-1 capital')
('NAME', 'Sevastopol')
('NAMEPAR', None)
('NAMEALT', None)
('NAMEASCII', 'Sevastapol')
('ADM0CAP', 0)
('CAPIN', None)
('WORLDCITY', 0)
('MEGACITY', 0)
('SOV0NAME', 'Russia')
('SOV_A3', 'RUS')
('ADM0NAME', 'Russia')
('ADM0_A3', 'RUS')
('ADM1NAME', 'Crimea')
('ISO_A2', 'RU')
('NOTE', None)
('LATITUDE', 44.599977)
('LONGITUDE', 33.464975)
('POP_MAX', 379200)
('POP_MIN', 314465)
('POP_OTHER', 290177.0)
('RANK_MAX', 10)
('RANK_MIN', 10)
('MEGANAME', None)
('LS_NAME', 'Sevastapol')
('MAX_POP10', 314465.0)
('MAX_POP20', 314465.0)
('MAX_POP50', 314465.0)
('MAX_POP300', 314465.0)
('MAX_POP310', 0.0)
('MAX_NATSCA', 100.0)
('MIN_AREAKM', 105.0)
('MAX_AREAKM', 105.0)
('MIN_AREAMI', 40.0)
('MAX_AREAMI', 40.0)
('MIN_PERKM', 111.0)
('MAX_PERKM', 111.0)
('MIN_PERMI', 69.0)
('MAX_PERMI', 69.0)
('MIN_BBXMIN', 33.425)
('MAX_BBXMIN', 33.425)
('MIN_BBXMAX', 33.633333)
('MAX_BBXMAX', 33.633333)
('MIN_BBYMIN', 44.533333)
('MAX_BBYMIN', 44.533333)
('MIN_BBYMAX', 44.675)
('MAX_BBYMAX', 44.675)
('MEAN_BBXC', 33.532042)
('MEAN_BBYC', 44.599245)
('TIMEZONE', 'Europe/Simferopol')
('UN_FID', 0.0)
('POP1950', 0.0)
('POP1955', 0.0)
('POP1960', 0.0)
('POP1965', 0.0)
('POP1970', 0.0)
('POP1975', 0.0)
('POP1980', 0.0)
('POP1985', 0.0)
('POP1990', 0.0)
('POP1995', 0.0)
('POP2000', 0.0)
('POP2005', 0.0)
('POP2010', 0.0)
('POP2015', 0.0)
('POP2020', 0.0)
('POP2025', 0.0)
('POP2050', 0.0)
('MIN_ZOOM', 4.0)
('WIKIDATAID', 'Q7525')
('WOF_ID', 101752787)
('CAPALT', 0)
('NAME_EN', 'Sevastopol')
('NAME_DE', 'Sewastopol')
('NAME_ES', 'Sebastopol')
('NAME_FR', 'Sébastopol')
('NAME_PT', 'Sebastopol')
('NAME_RU', 'Севастополь')
('NAME_ZH', '塞瓦斯托波尔')
('LABEL', None)
('NAME_AR', 'سيفاستوبول')
('NAME_BN', 'সেভাস্তোপোল')
('NAME_EL', 'Σεβαστούπολη')
('NAME_HI', 'सेवस्तोपोल')
('NAME_HU', 'Szevasztopol')
('NAME_ID', 'Sevastopol')
('NAME_IT', 'Sebastopoli')
('NAME_JA', 'セヴァストポリ')
('NAME_KO', '세바스토폴')
('NAME_NL', 'Sebastopol')
('NAME_PL', 'Sewastopol')
('NAME_SV', 'Sevastopol')
('NAME_TR', 'Sivastopol')
('NAME_VI', 'Sevastopol')
('NE_ID', 1159150561)
('NAME_FA', 'سواستوپول')
('NAME_HE', 'סבסטופול')
('NAME_UK', 'Севастополь')
('NAME_UR', 'سواستوپول')
('NAME_ZHT', '塞瓦斯托波')
('GEONAMESID', 694423.0)
('FCLASS_ISO', None)
('FCLASS_US', None)
('FCLASS_FR', None)
('FCLASS_RU', None)
('FCLASS_ES', None)
('FCLASS_CN', None)
('FCLASS_TW', None)
('FCLASS_IN', None)
('FCLASS_NP', None)
('FCLASS_PK', None)
('FCLASS_DE', None)
('FCLASS_GB', None)
('FCLASS_BR', None)
('FCLASS_IL', None)
('FCLASS_PS', None)
('FCLASS_SA', None)
('FCLASS_EG', None)
('FCLASS_MA', None)
('FCLASS_PT', None)
('FCLASS_AR', None)
('FCLASS_JP', None)
('FCLASS_KO', None)
('FCLASS_VN', None)
('FCLASS_TR', None)
('FCLASS_ID', None)
('FCLASS_PL', None)
('FCLASS_GR', None)
('FCLASS_IT', None)
('FCLASS_NL', None)
('FCLASS_SE', None)
('FCLASS_BD', None)
('FCLASS_UA', None)
('FCLASS_TLC', None)
('geometry', <POINT (33.524 44.604)>)

Not a damned one mentions Ukraine in any form.  The best I can do is to have an intermediary dereferencing via the states/provinces map data.   Here's every field in *that*:

('featurecla', 'Admin-1 states provinces')
('scalerank', 5)
('adm1_code', 'RUS-283')
('diss_me', 283)
('iso_3166_2', 'UA-43')
('wikipedia', None)
('iso_a2', 'RU')
('adm0_sr', 1)
('name', 'Crimea')
('name_alt', 'Crimée|Criméia|Krim|Krymskaya Respublika|Respublika Krym')
('name_local', None)
('type', 'Oblast')
('type_en', 'Autonomous Republic')
('code_local', None)
('code_hasc', 'UA.KR')
('note', None)
('hasc_maybe', 'UA.MY|UKR-KRN')
('region', 'Volga')
('region_cod', None)
('provnum_ne', 3.0)
('gadm_level', 1)
('check_me', 20)
('datarank', 5)
('abbrev', None)
('postal', 'KR')
('area_sqkm', 0.0)
('sameascity', -99)
('labelrank', 6)
('name_len', 6)
('mapcolor9', 6)
('mapcolor13', 3)
('fips', 'UP11')
('fips_alt', 'UP16')
('woe_id', 2347544.0)
('woe_label', 'Crimea, UA, Ukraine')
('woe_name', 'Crimea')
('latitude', 45.3115)
('longitude', 34.2784)
('sov_a3', 'RUS')
('adm0_a3', 'RUS')
('adm0_label', 2)
('admin', 'Russia')
('geonunit', 'Russia')
('gu_a3', 'RUS')
('gn_id', 703883.0)
('gn_name', 'Avtonomna Respublika Krym')
('gns_id', -1043929.0)
('gns_name', 'Krym, Avtonomna Respublika')
('gn_level', 1.0)
('gn_region', None)
('gn_a1_code', 'UA.11')
('region_sub', 'North Caucasus')
('sub_code', None)
('gns_level', 1.0)
('gns_lang', 'rus')
('gns_adm1', 'UP11')
('gns_region', None)
('min_label', 5.0)
('max_label', 11.0)
('min_zoom', 4.7)
('wikidataid', 'Q756294')
('name_ar', 'جمهورية القرم ذاتية الحكم')
('name_bn', 'স্বায়ত্তশাসিত প্রজাতন্ত্রী ক্রিমিয়া')
('name_de', 'Autonome Republik Krim')
('name_en', 'Autonomous Republic of Crimea')
('name_es', 'República autónoma de Crimea')
('name_fr', 'République autonome de Crimée')
('name_el', 'Αυτόνομη Δημοκρατία της Κριμαίας')
('name_hi', 'क्रीमिया')
('name_hu', 'Krími Autonóm Köztársaság')
('name_id', 'Republik Otonom Krimea')
('name_it', 'Repubblica autonoma di Crimea')
('name_ja', 'クリミア自治共和国')
('name_ko', '크림 자치 공화국')
('name_nl', 'Autonome Republiek van de Krim')
('name_pl', 'Republika Autonomiczna Krymu')
('name_pt', 'República Autónoma da Crimeia')
('name_ru', 'Автономная Республика Крым')
('name_sv', 'Autonoma republiken Krim')
('name_tr', 'Kırım Özerk Cumhuriyeti')
('name_vi', 'Cộng hòa Tự trị Krym')
('name_zh', '克里米亚自治共和国')
('ne_id', 1159309535)
('name_he', 'הרפובליקה האוטונומית של קרים')
('name_uk', 'Автономна Республіка Крим')
('name_ur', 'خود مختار جمہوریہ کریمیا')
('name_fa', 'جمهوری خودمختار کریمه')
('name_zht', '克里米亚自治共和国')
('FCLASS_ISO', None)
('FCLASS_US', None)
('FCLASS_FR', None)
('FCLASS_RU', None)
('FCLASS_ES', None)
('FCLASS_CN', None)
('FCLASS_TW', None)
('FCLASS_IN', None)
('FCLASS_NP', None)
('FCLASS_PK', None)
('FCLASS_DE', None)
('FCLASS_GB', None)
('FCLASS_BR', None)
('FCLASS_IL', None)
('FCLASS_PS', None)
('FCLASS_SA', None)
('FCLASS_EG', None)
('FCLASS_MA', None)
('FCLASS_PT', None)
('FCLASS_AR', None)
('FCLASS_JP', None)
('FCLASS_KO', None)
('FCLASS_VN', None)
('FCLASS_TR', None)
('FCLASS_ID', None)
('FCLASS_PL', None)
('FCLASS_GR', None)
('FCLASS_IT', None)
('FCLASS_NL', None)
('FCLASS_SE', None)
('FCLASS_BD', None)
('FCLASS_UA', None)
('FCLASS_TLC', None)
('geometry', <POLYGON ((33.629 46.125, 33.66 46.148, 33.671 46.173, 33.662 46.21, 33.67 4...>)

Then I can get the name from woe_label, but I have to do split on it and take the last element and strip spaces, EXCEPT that this isn't valid with all woe_element entries, so I have to have a fallback means to get the country through searching for the first half of the iso_3166_2 in *yet another* map, the sovereign states map.  And there are some things, like maps with internal lakes, for which there is *no* solution available.

This situation is beyond absurd - how many people are going to go through this much effort.   Every international organization recognizes Crimea as part of Ukraine, but Natural Earth has decided that it wants to implement an entirely different standard to make it part of Russia.  And - let me be clear, and given how Kelso refuses to engage on this - the Tenet Media indictment should raise **massive** red flags over any such behavior from a developer.

**Comment by qis** [comment]
@enn-nafnlaus Please don't randomly post in closed issues for no reason. I also encourage you to drop this topic altogether. The only ones who truly care about it, are people living in the Donbas, Novorossia, Crimea and Odessa regions and they are either very happy with their liberation by Russia or are eagerly awaiting it.

**Comment by goliney** [comment]
Let’s mark the Kursk region as a part of Ukraine then.

**Comment by qis** [comment]
That's an incursion. Only Ukrainian military forces are present there and have only killed and kidnapped some civilians, not established government structures, like Russia did in the liberated regions.

**Comment by goliney** [comment]
@qis This has been debunked as a russian false flag operation.

**Comment by enn-nafnlaus** [comment]
> @enn-nafnlaus Please don't randomly post in closed issues for no reason. I also encourage you to drop this topic altogether. The only ones who truly care about it, are people living in the Donbas, Novorossia, Crimea and Odessa regions and they are either very happy with their liberation by Russia or are eagerly awaiting it.

I literally have a friend from Mariupol you arse. The Russians killed both his brother and his wife, and obliterated his hometown, and he was shot in the abdomen while kicking them out of the Kherson area, but no, here you are to tell us how residents are secretly thrilled about it, and constantly lie in polls (eastern Ukraine is the region *most* against territorial concessions to Russia in polls).  Get the heck out of here with that.

And an "incursion" means a brief raid, not "taking more territory than Russia took all year in Ukraine, _Alexej_, and bringing in engineering equipment to fortify it".  And armed conscripts tasked to defend the border are not "civilians" (and are being almost immediately traded for Ukrainian POWs, who generally come back emaciated and with signs of torture).  If you want to talk about "abducting civilians", return the 20000 kidnapped children. Putin literally has a warrant out for his arrest from the International Criminal Court over this.

Let me be abundantly clear: the international community DOES NOT recognize Russia's invasion.  You can believe whatever the heck you want. You can believe that Hitler's invasion was right too if you want.  That doesn't make it so.
